Add reactivity control to coupled transport-depletion analyses

church89 opened the pull request:
# Description

This pr adds the capability to perform batch-wise reactivity control during a coupled transport-depletion analysis. 
It effectively runs an iterative `search_for_keff` sub-step algorithm to maintain keff equal to 1 (or some other user defined targets)  as a function of some user defined model parameter to update, before calculating the BOS from the transport run and proceed with the depletion step.
The model variable to parametrize during the sub-step can be of geometrical (e.g. control rod position) or material (e.g. refuel) nature, acting on a `openmc.Cell` and `openmc.Material`, respectively. 

A batch-wise scheme can be added to an `integrator` instance and a coupled transport-depletion simulation can be run like this:
```Python
...
...
op = openmc.deplete.CoupledOperator(model)
integrator = openmc.deplete.CECMIntegrator(op, timesteps , power)
integrator.add_batchwise(cell, 'translation', axis = 2, bracket = [-4, 4], bracket_limit = [-100,20])
integrator.integrate()
```
where `cell` is an `openmc.Cell` conveniently defined  that can be translated along the `z-axis` during the run. 




# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
<!--
While tests will automatically be checked by CI, it is good practice to
ensure that they pass locally first. See instructions here:
https://docs.openmc.org/en/latest/devguide/tests.html
-->

church89 committed:
first commit to new branch

church89 committed:
first commit to new branch

church89 committed:
batchwise first commits

church89 committed:
update unit test

church89 committed:
update unit test

church89 committed:
add test unit

church89 committed:
add test unit

church89 committed:
unit test update

church89 committed:
unit test update

church89 committed:
further improvements to the module

church89 committed:
further improvements to the module

church89 committed:
add unit and regression tests

church89 committed:
add unit and regression tests

church89 committed:
Merge pull request #10 from openmc-dev/develop

update pr with latest commits from development branch

church89 committed:
add missing docstrings

church89 committed:
add missing docstrings

church89 committed:
remove unused import

church89 committed:
remove unused import

church89 committed:
add missing returns

church89 committed:
add missing returns

church89 committed:
fix typo

church89 committed:
fix typo

church89 committed:
Merge pull request #11 from openmc-dev/develop

Update with latest develop before open pr

church89 committed:
prevent to openmc batchwise_root when to there

church89 committed:
prevent to openmc batchwise_root when to there

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+    target : Real, optional
+        This is equivalent to the `target` parameter of the `search_for_keff`.
+        Default to 1.0.
+    print_iterations : Bool, Optional
+        Whether or not to print `search_for_keff` iterations.
+        Default to True
+    search_for_keff_output : Bool, Optional
+        Whether or not to print transport iterations during  `search_for_keff`.
+        Default to False
```
These variables seem very similar and could use some updated docs to explain what they're doing. Is this sufficient and correct?
```suggestion
    print_iterations : Bool, Optional
        Print a status message each iteration
        Default to True
    search_for_keff_output : Bool, Optional
        Print full transport logs during iterations (e.g., inactive generations, active
        generations). Default to False
```

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+                         bracket_limit[0], bracket_limit[1])
+        self.bracket_limit = bracket_limit
+
+        check_value('bracketed_method', bracketed_method,
+                   _SCALAR_BRACKETED_METHODS)
+        self.bracketed_method = bracketed_method
```
Redundant `check_value` here? We check already in the setter, so can we just do `self.bracketed_method = bracketed_method`

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+        self.model = model
+        self.geometry = model.geometry
+
+        check_value('density_treatment', density_treatment,
+                                        ('constant-density','constant-volume'))
+        self.density_treatment = density_treatment
+
+        check_iterable_type('bracket', bracket, Real)
+        check_length('bracket', bracket, 2)
+        check_less_than('bracket values', bracket[0], bracket[1])
+        self.bracket = bracket
+
+        check_iterable_type('bracket_limit', bracket_limit, Real)
+        check_length('bracket_limit', bracket_limit, 2)
+        check_less_than('bracket limit values',
+                         bracket_limit[0], bracket_limit[1])
+        self.bracket_limit = bracket_limit
+
+        check_value('bracketed_method', bracketed_method,
+                   _SCALAR_BRACKETED_METHODS)
+        self.bracketed_method = bracketed_method
+
+        check_type('tol', tol, Real)
+        self.tol = tol
+
+        check_type('target', target, Real)
+        self.target = target
+
```
It feels a bit redundant to have `check*` calls in the `__init__` and then similar checks in the property setters. I would prefer having all the checks in the setters, so that way we have a smaller place to change things / add checks

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+    search_for_keff_output : Bool, Optional
+        Whether or not to print transport iterations during  `search_for_keff`.
+        Default to False
+    Attributes
```
Please reevaluate your docstrings to adhere to the [`numpydoc` spec](https://numpydoc.readthedocs.io/en/latest/format.html#sections), especially in regards to empty lines between these various settings

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+        #make sure we don't modify original bracket and tol values
+        bracket = deepcopy(self.bracket)
+
+        #search_for_keff tolerance should vary according to the first guess value
+        if abs(val) > 1.0:
+            tol = self.tol / abs(val)
+        else:
+            tol = self.tol
```
Why does tolerance need to vary according to the first value? Is it supposed to be relative but it's given as absolute? Or vice versa. Or is `val` not a real number the first iteration?

Please consider removing this conversion

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+        # Run until a search_for_keff root is found or out of limits
+        root = None
+
+        while root == None:
+            search = search_for_keff(self._model_builder,
+                            bracket = np.array(bracket) + val,
+                            tol = tol,
+                            bracketed_method = self.bracketed_method,
+                            target = self.target,
+                            print_iterations = self.print_iterations,
+                            run_args = {'output': self.search_for_keff_output},
+                            run_in_memory = True)
```
I get nervous when I see a `while` loop that doesn't have a clear exit strategy. Consider the case where I want to run a critical boron search through depletion, but later in the simulation I've burned so much of my fuel that even with no boron I am subcritical. This loop would run forever.

Can we have the user provide an iteration limit, even defaulting to a high number, so we have a guaranteed exit condition?

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+    @bracketed_method.setter
+    def bracketed_method(self, value):
+        check_value('bracketed_method', value, _SCALAR_BRACKETED_METHODS)
+        if value != 'brentq':
+            warn('WARNING: brentq bracketed method is recommended')
```
Can we instead put this in some documentation somewhere? If other methods are not recommended, why do we support them? Sure it's good to have more options, and there is usually a good case to be made for testing out new algorithms. 

I feel a better user experience would be to have some documentation indicating the default `brentq` method is recommended and this warning be removed.

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+        val : openmc.Cell
+            Openmc Cell
+        """
+        if isinstance(val, Cell):
+            check_value('Cell exists', val, [cell for cell in \
+                                self.geometry.get_all_cells().values()])
+
+        elif isinstance(val, str):
+            if val.isnumeric():
+                check_value('Cell id exists', int(val), [cell.id for cell in \
+                                self.geometry.get_all_cells().values()])
+                val = [cell for cell in \
+                       self.geometry.get_all_cells().values() \
+                       if cell.id == int(val)][0]
+            else:
+                check_value('Cell name exists', val, [cell.name for cell in \
+                                self.geometry.get_all_cells().values()])
+
+                val = [cell for cell in \
+                       self.geometry.get_all_cells().values() \
+                       if cell.name == val][0]
+
+        elif isinstance(val, int):
+            check_value('Cell id exists', val, [cell.id for cell in \
+                                self.geometry.get_all_cells().values()])
+            val = [cell for cell in \
+                       self.geometry.get_all_cells().values() \
+                       if cell.id == val][0]
+        else:
```
That's a lot of calls to `self.geometry.get_all_cells` which could be super expensive for large models. Can we instead store all the potential cells at the top of this method?

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+        val = self._get_cell_attrib()
+        check_type('Cell coeff', val, Real)
+
+        # Update all material densities from concentration vectors
+        #before performing the search_for_keff. This is needed before running
+        # the  transport equations in the search_for_keff algorithm
+        super()._update_materials(x)
+
+        # Calculate new cell attribute
+        root = super()._search_for_keff(val)
```
Nit pick but we don't usually see calls to `super` when we're calling a different method from the parent class. It shows up in `__init__` because we want to call the parent classes `__init__` method, but we don't need it all the time. Recommend replacing these `super` calls with `self._update_materials` and `self._search_for_keff`

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+        # new volume
+        if self.cell_materials:
+            volumes = self._calculate_volumes()
+            x = super()._update_x_and_set_volumes(x, volumes)
```
Same note on the calls to `super`

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+            Total atoms concentrations
+        Returns
+        -------
+        x : list of numpy.ndarray
```
We're returning two items here, but only one is documented. Please update

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1157 @@
+        coeff : float
+            cell coefficient
+        """
+        for cell in openmc.lib.cells.values():
+            if cell.id == self.cell.id:
+                if self.attrib_name == 'translation':
+                    return cell.translation[self.axis]
+                elif self.attrib_name == 'rotation':
+                    return cell.rotation[self.axis]
```
For these calls when we're digging through the cells in the model, can we instead save the desired cell instance on this class? Maybe that isn't super ideal because we're something from the c++ side. But that's an awful lot of iterations over cells when a simple `self._cell` seems cheap and fast

drewejohnson commented on tests/regression_tests/deplete_with_batchwise/test.py:
```diff
@@ -0,0 +1,122 @@
+
+    return openmc.Model(geometry, materials, settings)
+
+@pytest.mark.skipif(sys.version_info < (3, 9), reason="Requires Python 3.9+")
```
Why does this need `python 3.9+`? Our [python guide](https://docs.openmc.org/en/stable/devguide/styleguide.html#python) (which maybe needs to be updated) says things should work on python 3.7 (now obselete to maybe a larger discussion)

drewejohnson replied:
Please address this change with an intention to remove the restriction on python >= 3.9

drewejohnson commented on tests/regression_tests/deplete_with_batchwise/test.py:
```diff
@@ -0,0 +1,122 @@
+    res_test = openmc.deplete.Results(path_test)
+    res_ref = openmc.deplete.Results(path_reference)
+
+    # Use high tolerance here
```
Oof a 2% relative tolerance is high. But we've got a stochastic code, and probably wanting to get the simulation to converge reasonably quickly.

I'd like to see a tighter tolerance but I understand that might not be tenable.

drewejohnson reviewed (changes requested):
@church89 thank you for this lengthy and useful PR! I think this is going to unlock a lot of doors for people and provide a lot of value. I haven't done a full review but I wanted to get some of my larger thoughts out first, make sure we understand the vision and goal, and then refine my review as we go.

Some simple things
- Please re-evaluate the docstrings you've added to align with the [`numpydoc` spec](https://numpydoc.readthedocs.io/en/latest/format.html#sections) Otherwise our built documentation isn't going to process this properly
- Please run a python formatter like `black` to make sure your code adheres to [pep8 formatting style](https://www.python.org/dev/peps/pep-0008/) like the [dev guide requests](https://docs.openmc.org/en/stable/devguide/styleguide.html#python)
- Please consolidate property setters / getters / validators. I think it's going to be easier for us in the long run to maintain / add validators if we do validation on things like `self.tol` if all validation is done in property setters

Maybe a larger lift: I think the high-level class and method name, using batch wise, might not communicate what we want to communicate. My rationale is I believe this engages after a full transport simulation, while `openmc` uses the "batch" terminology to mean generations of particles in a given transport simulation - [docs](https://docs.openmc.org/en/stable/usersguide/settings.html?highlight=batch#number-of-generations-per-batch)

so users who already are configuring an openmc run to have so many inactive / active generations per *batch* may expect this search to be performed more often than it is actually performed. 

But, based on your PR message

> It effectively runs an iterative search_for_keff sub-step algorithm to maintain keff equal to 1 (or some other user defined targets) as a function of some user defined model parameter to update, before calculating the BOS from the transport run

and my non-exhaustive review, I could be wrong. If we are making changes to the model *during* batches of active generations, then this name makes sense. But if we are not, I would encourage us to consider new names.

I hope to have time in the next week or so for a more detailed review, but I wanted to get something out the door to let you know I've seen this, I believe it's valuable, and I have some (hopefully) low-hanging requests

church89 committed:
Apply suggestions from code review

Co-authored-by: Andrew Johnson <drewejohnson@users.noreply.github.com>

church89 committed:
Apply suggestions from code review

Co-authored-by: Andrew Johnson <drewejohnson@users.noreply.github.com>

church89 commented on openmc/deplete/abc.py:
```diff
@@ -847,6 +869,33 @@ def add_transfer_rate(self, material, components, transfer_rate,
         self.transfer_rates.set_transfer_rate(material, components, transfer_rate,
                                       transfer_rate_units, destination_material)
 
+    def add_batchwise(self, obj, attr, **kwargs):
+        """Add batchwise operation to integrator scheme.
+
+        Parameters
+        ----------
+        obj : openmc.Cell or openmc.Material object or id or str name
+            Cell or Materials identifier to where add batchwise scheme
+        attr : str
+            Attribute to specify the type of batchwise scheme. Accepted values
+            are: 'translation', 'rotation', 'temperature' for an openmc.Cell
+            object; 'refuel' for an openmc.Material object.
+        **kwargs
+            keyword arguments that are passed to the batchwise class.
+
+        """
+        check_value('attribute', attr, ('translation', 'rotation',
+                                        'temperature', 'refuel'))
+        if attr in ('translation', 'rotation'):
+            batchwise = BatchwiseCellGeometrical
+        elif attr == 'temperature':
+            batchwise = BatchwiseCellTemperature
+        elif attr == 'refuel':
+            batchwise = BatchwiseMaterialRefuel
+
+        self.batchwise = batchwise.from_params(obj, attr, self.operator,
+                                           self.operator.model, **kwargs)
+
```
yes, I've tried to generalize as much as possible these cases here, but of course there will be limitations and a user may want to define a particular case not covered here. In the long term I would like implement something that can give such an option. 

I agree that `self.batchwise` should be protected, but I also like the `operator.add_batchwise` method, which makes it pretty neat. For now I'll just change `self.batchwise` into  `self._batchwise`, but we can come back to this point. 

Indeed, `self.operator.model` argument is not needed, thanks :)

church89 committed:
address comments by @drewejohnson

church89 committed:
address comments by @drewejohnson

church89 commented:
Hi @drewejohnson, thanks for starting this review. I've begun to address your comments/suggestions.
 
I can see how `batchwise` might create some confusion. The routine acts in between a transport and a depletion step, so an alternatives could be `stepwise` to maintain the general purpose. 
Otherwise, we might think of linking the concepts of depletion and criticality, as ultimately is what it does, like `depletion_critical` , `criticality_depletion` or something similar.

church89 committed:
fix some docstrings

church89 committed:
fix some docstrings

church89 committed:
reformatted with black

church89 committed:
reformatted with black

church89 committed:
refactoring of _search_for_keff method

church89 committed:
refactoring of _search_for_keff method

church89 committed:
fix some spelling typos

church89 committed:
fix some spelling typos

church89 committed:
resolve conflicts

church89 committed:
fix conflicts

church89 committed:
missing underscore

church89 committed:
add batchwise get method and update unit test

drewejohnson commented on openmc/deplete/abc.py:
```diff
@@ -763,6 +773,14 @@ def _get_start_data(self):
+        """
+        x = deepcopy(bos_conc)
+        # Get new vector after keff criticality control
+        x, root = self._batchwise.search_for_keff(x, step_index)
```
Subclasses of the abstract base class are not required to implement `search_for_keff` since it is not defined on the base class at all, let alone an abstract method. Please add a corresponding abstract `search_for_keff(...)` method.

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1316 @@
+        # even though they are also re-calculated when running the search_for_kef
+        self._update_materials(x)
+
+        # Solve search_for_keff and find new value
+        root = self._search_for_keff(0)
```
A comment would be useful as to why we are passing a value of `0` here. Because we are starting at a point of making no modifications?

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1316 @@
+class BatchwiseMaterialRefuel(BatchwiseMaterial):
+    """Batch wise material-based class for refuelling (addition or removal) scheme.
+
+    A user doesn't need to call this class directly.
+    Instead an instance of this class is automatically created by calling
+    :meth:`openmc.deplete.Integrator.add_batchwise` method from an integrator
+    class, such as  :class:`openmc.deplete.CECMIntegrator`.
+
+    .. versionadded:: 0.13.4
```
Same comment on the version added string and allowing the user to construct these directly

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1316 @@
+        self.attrib_name = attrib_name
+
+    def _model_builder(self, param):
+        """Callable function which builds a model according to a passed
```
Please ensure docstrings have a single line and then an empty line before new content

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1316 @@
+        return self.model
+
+    def _calculate_volumes(self, res):
+        """Uses :meth:`openmc.batchwise._search_for_keff` solution as grams of
```
Please ensure docstrings have a single line and then an empty line before new content

drewejohnson commented on openmc/deplete/stepresult.py:
```diff
@@ -74,7 +74,7 @@ def __init__(self):
 
         self.data = None
-
+        self.batchwise = None
```
New attribute should be documented like other attributes

drewejohnson commented on openmc/deplete/stepresult.py:
```diff
@@ -509,7 +523,7 @@ def from_hdf5(cls, handle, step):
 
     @staticmethod
     def save(op, x, op_results, t, source_rate, step_ind, proc_time=None,
-             path: PathLike = "depletion_results.h5"):
+             root=None, path: PathLike = "depletion_results.h5"):
```
New argument to public method should be documented

drewejohnson commented on tests/unit_tests/test_deplete_batchwise.py:
```diff
@@ -0,0 +1,204 @@
+    """
+    Test classes attributes are set correctly
+    """
+    op = CoupledOperator(model, CHAIN_PATH)
+
+    int = openmc.deplete.PredictorIntegrator(
+        op, [1,1], 0.0, timestep_units = 'd')
```
This integrator is used a lot in this file. Could we make it a `pytest.fixture` instead?

drewejohnson reviewed (changes requested):
I am leaving a review of the unit tests because I think the changes to the implementation file should be addressed first. Those could change some of the tests so I'll circle back to those.

Following up from the discussion on calling this `batchwise`, I think `reactivity_control` makes a lot of sense. If you agree, please make at least the following changes

1. `openmc.deplete.batchwise` -> `openmc.deplete.reactivity_control`
2. `Batchwise` base class to `ReactivityController`
3. `BatchwiseCell` base class to `CellReactivityController`
4. `BatchwiseCellGeometrical` to `GeometricalCellReactivityController`
5. and similar changes to the other classes, variables, attributes, etc. that use the batchwise naming.

This feels like a lot of functionality and math to only _really_ have covered by one regression test. There are some good test on the "modifier" methods (cell rotation, translation, refueling) but it's hard to tell if the core search is really working as intended without the regression test. Surely there are some missed branches of code depending on if the search was successful or not, if the mean keff was above or below the target, if the last keff was greater than the first, etc. I would love to see some testing on some of the internals of the adaptive search routine to help convince me and future devs that this is doing what we expect. And to make sure we don't accidentally introduce some bugs down the road!

This could be daunting because the "core" `_search_for_keff` routine is long and complex. I would recommend trying to find small logically similar operations and move those to additional helper methods on the base class. If they're static methods, we don't need to worry about constructing a concrete version of the base class either.

One example would be the adaptive bracket search. That could be extracted to a staticmethod or function that looked like
```python
def adjust_brackets(brackets, guesses, keffs):
```
that returned the new brackets given previous guesses and keffs. Then, multiple unit tests can be added that poke and prod this code from different angles to make sure it's doing what we want.

church89 committed:
Apply suggestions from code review

Address doc string comments

Co-authored-by: Andrew Johnson <drewejohnson@users.noreply.github.com>

church89 committed:
address comments by @drewejohnson

church89 committed:
address more comments by @drewejohnson

church89 committed:
replaced batchwise nomenclature with reactivity_control one

church89 committed:
update regression reference result files

church89 commented:
@drewejohnson thanks, I've addressed most of your comments and changed the name to `reactivity_control`, which I like it and think also is appropriate. 
Regarding the adaptive search routine, as I tried to explain in one of the review comment, is probably less complicated than it looks, and right now relies on the fact that the method for `search_for_keff` is a bracketed one, so that convergence within tolerance is always ensured and we can simply rely on the algorithm to move the bracket to where the two limits have opposite sign with respect to the target.

church89 committed:
few fixes

church89 committed:
fix unit test failing

church89 committed:
explicit path argument in SIIntegrator step result save method, since right thi class is not supported

church89 committed:
add reactivity control restarting condition to integrator class

church89 committed:
Merge branch 'develop' into batchwise_pr

church89 committed:
remove unused index argumnet in restart method to comply with latest develop branch

church89 committed:
make sure cell attribute vector maintain initial coordinates

church89 committed:
make sure bracket type is an array and apply black code style

church89 committed:
fix failing regression test

church89 committed:
update reference result files

church89 committed:
change warning message

church89 committed:
restore check conditions in adaptive search_for_keff routine

drewejohnson commented on openmc/search.py:
```diff
@@ -51,7 +51,11 @@ def _search_keff(guess, target, model_builder, model_args, print_iterations,
     model = model_builder(guess, **model_args)
 
     # Run the model and obtain keff
-    sp_filepath = model.run(**run_args)
+    if run_in_memory:
+        openmc.lib.run(**run_args)
+        sp_filepath = f'statepoint.{model.settings.batches}.h5'
+    else:
+        sp_filepath = model.run(**run_args)
```
@paulromano @pshriwise or whomever, can I get another opinion on these changes to `openmc/search.py`?

paulromano replied:
Yes, ideally this could just be its own (much smaller) PR!

drewejohnson commented on openmc/search.py:
```diff
@@ -193,12 +197,26 @@ def search_for_keff(model_builder, initial_guess=None, target=1.0,
+        # In case the root finder is not successful
+        except Exception as e:
+            print(f'WARNING: {e}')
+            return guesses, results
```
@church89 I'm interested in getting feedback from other people on these changes. My personal preference is having a consistent return signature (always three values) rather than sometimes two, sometime three.

But, moving to returning three values is a large API change that would benefit from discussion with other users and developers.

The decision needs to also be reflected in the documentation for this function: when three values are returned or not, or adding the `zero_value` to the return block

ping @paulromano @pshriwise @shimwell or whomever else

drewejohnson commented on openmc/deplete/batchwise.py:
```diff
@@ -0,0 +1,1316 @@
+
+        # Run until a search_for_keff root is found or out of limits
+        root = None
+        while root == None:
```
@church89 please consider incorporating this request for a user provided iteration limit, even defaulting to a large positive number. Especially for depletion problems, it's possible you get to a case where you cannot add or remove enough reactivity to counterbalance the consumption of fuel

drewejohnson reviewed (changes requested):
Thank you for incorporating some of the changes. I want to make one last call for the following items to be either incorporated or rejected with explanation.

## Infinite loop
I've personally been bitten by an unterminated while loop and it can be deeply frustrating. And, since users of this feature are likely going to be running expensive simulations, there is a likelihood that someone expends some unnecessary computational resources because there is no hard upper limit on iterations.

If there is some other way that I'm not seeing where the main loop will terminate if it cannot find a satisfactory solution, please let me know.

## reactivity controller setter

The `add_reactivity_control` method is nice, I'm not trying to deny that. But it does not make it easy for people to develop their own reactivity controller outside of openmc and then use that in a simulation. They would need to modify the source code (either locally or through a PR here) to add their specific flavor of controller. Or they push an issue here and wait for someone else to make those changes.

If the `reactivity_controller` is a publicly settable attribute, then people have the freedom to use their own controller without needing to make changes to the core of openmc. It can (and should) be used by the `add_reactivity_control` helper method.

## search.py

These changes don't feel intrinsically related to the PR at hand. They feel valuable, but the relationship is not clear to me. There is some issue you are trying to workaround that isn't transparent to me.

I also have some concerns about variable return signature (two or three arguments) and would welcome other opinions on the matter.

church89 commented:
@drewejohnson 
Thank you for your comments. Regarding your concerns:

- **Infinite loop**: As long as the `search_for_keff` bracketed algorithm returns two arguments (meaning that the search have failed to bound the root) the adaptive algorithm will try to move the bracket range to bound the target. When that happens the `search_for_keff` returns three arguments and the first exit point to the while loop. 
If the adaptive algorithm moves the bracket range outside of the user defined limits (e.g. the entire span of a control rod), the closest limit is set as the root and the while loop exits as well. 
The only case the loop won't exit easily is if the adaptive algorithm won't bound the target but still remain inside the user defined limits, this might take awhile depending on user settings and simulation model but eventually should make it. 
Inside the loop there is also another check condition that is: when `search_for_keff` returns two arguments, (`guesses` and `keffs`) and `keffs` difference is below 1 pcm, the algorithm is already very close to target and the root will be set as the closest guess to target, exiting the loop. This 1 pcm can be user defined, but one should not go lower than 1, otherwise there is a risk of overshooting the adaptive bracket out of user limits and invalid the method

- **Reactivity control setter**: Generally I like your idea and I think it makes sense, in practice I think it is not so straightforward and by letting users define their own controllers there is still a high risk of not being able to prevent misleading code. We would need to think of some very general helper methods that would need to include any user ideas and that might not be an easy task. What cases were you thinking that are not already covered by existing controllers? 

- **Search method**: These changes here are related to this PR. As explained above in the infinite loop, the `search_for_keff` must be able to fail (returning two arguments instead of three) to enable the adaptive algorithm and find the right bracket range the bounds the root. The `run_in_memory` argument instead is simply to enable `search_for_keff` to work with a model instantiated in memory. This is also related, as the operator class runs the model in memory.

church89 committed:
made ReactivityControllers settable through the integrator

church89 committed:
Merge pull request #31 from openmsr/reactivity_control

made ReactivityControllers settable through the integrator

church89 committed:
extend use of paramteric geometry to lattice and not only universe

drewejohnson commented on openmc/deplete/abc.py:
```diff
@@ -701,6 +712,33 @@ def solver(self, func):
+    @reactivity_control.setter
+    def reactivity_control(self, reactivity_control):
+        if isinstance(reactivity_control, CellReactivityController):
+            if not isinstance(reactivity_control.cell, Cell):
+                raise TypeError(
+                    "Reactivity control attribute must be Cell "
+                    "not {}".format(type(reactivity_control.cell)))
```
These feel like checks that would belong on the `CellReactivityController` class. Such that it would be impossible to create a cell reactivity controller where the `cell` attribute is not a `Cell`

drewejohnson commented on openmc/deplete/abc.py:
```diff
@@ -701,6 +712,33 @@ def solver(self, func):
+
+    @reactivity_control.setter
+    def reactivity_control(self, reactivity_control):
+        if isinstance(reactivity_control, CellReactivityController):
+            if not isinstance(reactivity_control.cell, Cell):
+                raise TypeError(
+                    "Reactivity control attribute must be Cell "
+                    "not {}".format(type(reactivity_control.cell)))
+        elif isinstance(reactivity_control, MaterialReactivityController):
+            if not isinstance(reactivity_control.material, Material):
+                raise TypeError(
+                    "Reactivity control attribute must be Material "
+                    "not {}".format(type(reactivity_control.material)))
+        else:
+            raise TypeError(
+                "Reactivity Control must either be "
+                "CellReactivityController, or MaterialReactivityController, "
+                "not {}".format(type(reactivity_control)))
```
What is the concern about only enforcing `isinstance(reactivity_control, ReactivityController)`? It's already an abstract base class so as long as someone fills out the abstract methods, any controller should work. Surely most cases will be modifying properties on cells or materials, but why forbid people from rolling their own `ReactivityController` subclass? If it's to set up protections and ensure the classes are well defined, what else can we put on the abstract base class to guide people? 

Otherwise, we would need to come back to this setter and make modifications to the source code to allow more extensions.

Recommend reading on the [Open/Closed principle of object oriented programming](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle)

> software entities (classes, modules, functions, etc.) should be open for extension, but closed for modification

drewejohnson commented on openmc/deplete/abc.py:
```diff
@@ -701,6 +712,33 @@ def solver(self, func):
+                "CellReactivityController, or MaterialReactivityController, "
+                "not {}".format(type(reactivity_control)))
+
+        check_value('attribute', reactivity_control.attrib_name,
+                    ('translation', 'rotation', 'temperature', 'refuel'))
```
Why only these four control mechanisms? What about the interplay between the `Operator` and the `ReactivityController` requires only these four mechanisms?

For light water reactors, adding and changing soluble boron concentration might be akin to "refueling" but we're not adding fuel there so someone could set `attrib_name = "dillution"` (for example).

drewejohnson commented on openmc/deplete/abc.py:
```diff
@@ -701,6 +712,33 @@ def solver(self, func):
+                    "Reactivity control attribute must be Cell "
+                    "not {}".format(type(reactivity_control.cell)))
+        elif isinstance(reactivity_control, MaterialReactivityController):
+            if not isinstance(reactivity_control.material, Material):
+                raise TypeError(
+                    "Reactivity control attribute must be Material "
+                    "not {}".format(type(reactivity_control.material)))
```
Same comment on ensuring that there is no way to construct a `MaterialReactivityController` such that `material` is not a `Material`.

This allows us to shift the responsibility to check the validity off the client (`Operator`) and onto the construction of the thing it's using.

drewejohnson reviewed (dismissed):
Apologies again for my delay. Life has been a bit hectic and I finally found time to review this.

Thank you for adding the reactivity control setter. I think it enforces too many checks and doesn't allow users to expand beyond what OpenMC can do already without source code changes. Please consider requiring the setter to only check that the object is an instance of the base `ReactivityController`. If there are concerns about allowing the users to construct those subclasses properly, what checks and validation steps can be added to the subclasses? e.g., to ensure that `CellReactivityController.cell` is always a `Cell` subclass?

Regarding the infinite loop,

> The only case the loop won't exit easily is if the adaptive algorithm won't bound the target but still remain inside the user defined limits

This is not an uncommon situation. Consider control rod movement through depletion. It is conceivable that, towards the end of life, there is no way to remove the control rods enough from the reactor to achieve a critical state with sufficiently spent fuel. Same for other reactivity control mechanisms where a poison is modified (control rods, control drums, soluble boron, etc.)

I am going to signal for an additional reviewer to help close the gap on the infinite loop and API changes to `openmc/search.py`.

church89 committed:
implement suggested changes to reactivity control classes

church89 committed:
update tests

church89 committed:
Merge pull request #32 from openmsr/reactivity_control_abstrac

Reactivity control abstrac

church89 commented:
Hi @drewejohnson and thanks again.

> Apologies again for my delay. Life has been a bit hectic and I finally found time to review this.
> 
> Thank you for adding the reactivity control setter. I think it enforces too many checks and doesn't allow users to expand beyond what OpenMC can do already without source code changes. Please consider requiring the setter to only check that the object is an instance of the base `ReactivityController`. If there are concerns about allowing the users to construct those subclasses properly, what checks and validation steps can be added to the subclasses? e.g., to ensure that `CellReactivityController.cell` is always a `Cell` subclass?
> 


I've reduced the reactivity controller classes to only two: `CellReactivityController` and `MaterialReactivityController`, which both inherit from the abstract `ReactivityController`, and update the setter.  It should be possible in principle now to have the users defined their own controllers. 

> Regarding the infinite loop,
> 
> > The only case the loop won't exit easily is if the adaptive algorithm won't bound the target but still remain inside the user defined limits
> 
> This is not an uncommon situation. Consider control rod movement through depletion. It is conceivable that, towards the end of life, there is no way to remove the control rods enough from the reactor to achieve a critical state with sufficiently spent fuel. Same for other reactivity control mechanisms where a poison is modified (control rods, control drums, soluble boron, etc.)
> 

That's exactly the reason we have to set a  `bracket_limit`: if one of the bounds (upper or lower) gets hit the **while loop** exits and the search is stopped. 
In case of a removable control rod, you would set this `bracket_limit` to be the active length of your rod inside the core, so that when the it gets pulled out completely the routine knows that there is not margin anymore and needs to stop.

church89 committed:
add forgotten argument to StepResult save method

church89 committed:
Merge remote-tracking branch 'openmsr/add_path_to_step_res' into batchwise_pr

church89 committed:
Merge branch 'develop' into batchwise_pr

church89 committed:
Merge pull request #44 from openmc-dev/develop

merge latest develop

paulromano reviewed (changes requested):
Finally getting around to giving this a proper review. Thanks for the work on this PR and for your patience @church89. And also many thanks to @drewejohnson for providing very thoughtful comments that have helped shaped the PR thus far.

Overall I have a few thoughts:
1. I personally like the idea of giving users more control over the criticality search and keeping it general rather than having a bunch of subclasses of a ReactivityController class. Right now, if a user wants to do some other search, they are forced to write a class that implements the ReactivityController interface. I feel like this could be much simpler overall. There could be a generic reactivity controller that accepts a function that takes one argument (the search value) and makes a corresonding change in the model via openmc.lib. So for example, if the user wanted to translate cell 50 along the y axis, the function could be as simple as:
    ```Python
    def translate_cell(value):
        cell = openmc.lib.cells[50]
        cell.translation = (0, value, 0)
    ```
    This may arguably be less fool-proof than providing _specific_ controllers but we can try to make things more obvious for users by providing examples (in the docstrings and/or example notebooks).
2. I agree with @drewejohnson's earlier comment about the changes in `search_for_keff`. Let's separate those out into a new PR to get them merged sooner.
3. I've been wanting to make some other changes in `search_for_keff`, namely to use the `scipy.optimize.root_scalar` function that can dispatch to `brentq`, `bisect`, etc. This would make it future-proof (if a new method is added, we don't need to update OpenMC). It probably makes sense to get that in before we merge this PR since it might affect the interface, so I'll try to get that PR submitted soon.

paulromano reviewed (changes requested):
@church89 Now that #3569 has been merged, can you update this PR based on the earlier discussion? It also looks like there are some conflicts that need to be resolved. Let me know if I can help at all!

church89 committed:
Merge branch 'develop' of https://github.com/openmc-dev/openmc into openmc-dev-develop

church89 committed:
Merge branch 'openmc-dev-develop' into batchwise_pr

church89 committed:
add lost docstring

church89 committed:
refactoring of reactivity control

church89 committed:
remove obsolete bits and update unit test

church89 committed:
minor fix

church89 committed:
minor fix

church89 committed:
Merge pull request #54 from openmsr/batchwise_pr_keff_search_update

Batchwise pr keff search update

church89 commented:
@paulromano this PR has finally been refactored and simplified a lot based on the awesome #3569 and earlier discussions. Haven't fully tested it yet but you can go through it when you have some time and tell me what you think!

paulromano commented on openmc/deplete/abc.py:
```diff
@@ -1048,6 +1086,113 @@ def add_redox(self, material, buffer, oxidation_states, timesteps=None):
+            x0,
+            x1,
+            bracket,
+            **kwargs)
```
`ReactivityController` takes `kwargs` as a single argument right now; it should be changed to take `**kwargs`

paulromano replied:
Also, I would rename this argument throughout to `search_kwargs` to make clear that it is keyword arguments being passed to the search function

paulromano commented on openmc/deplete/keff_search_control.py:
```diff
@@ -0,0 +1,145 @@
+            Updated atom density vector synchronized across all ranks
+        """
+        for rank in range(comm.size):
+            number_i = comm.bcast(self.operator.number, root=rank)
```
If the density change is being made through `openmc.lib`, wouldn't this end up resetting the densities to what they were before since `operator.number` doesn't get updated?

paulromano commented on openmc/deplete/abc.py:
```diff
@@ -877,10 +900,19 @@ def integrate(
+                    if self._reactivity_control:
+                        root = self.operator.prev_res[-1].reac_cont
+                    else:
+                        root = None
```
It doesn't look like anything is being done with the root in the case that it is present. Is the intention to reapply the stored root or should we re-run the keff search in the case of a restart?

church89 replied:
Ok, It was mainly coming from the old code. I guess here with the new implementation it is safe to just re-run e keff_search

paulromano commented on openmc/deplete/stepresult.py:
```diff
@@ -72,6 +72,7 @@ def __init__(self):
         self.mat_to_hdf5_ind = None
 
         self.data = None
+        self.reac_cont = None
```
I would probably name this `reac_cont_root` to be more explicit. It should also be described above in the docstring.

paulromano commented on openmc/deplete/reactivity_control.py:
```diff
@@ -0,0 +1,145 @@
+                f"keff search result ({root:.6f}) is above the upper bracket bound "
+                f"({self.kwargs['x_max']:.6f}). Clamping to bracket upper bound.",
+                UserWarning
+            )
```
It says it's being clamped but I don't actually see any clamping below?

paulromano commented on openmc/deplete/reactivity_control.py:
```diff
@@ -0,0 +1,145 @@
+        self.kwargs['x_max'] = bracket[1]
+        
+    def __call__(self, *args, **kwargs):
+        return self.function(*args, **kwargs)
```
This doesn't appear to be used?

paulromano commented on openmc/deplete/reactivity_control.py:
```diff
@@ -0,0 +1,145 @@
+import numpy as np
```
Unused

paulromano reviewed (changes requested):
Thanks for the updates on this PR @church89! Regarding the name of this feature, I think it would be more explicit to have:
- `Integrator.add_keff_search`
- `KeffSearchControl`

Also, since `KeffSearchControl` doesn't appear to be something users would instantiate directly, I think it should be an internal class (`_KeffSearchControl`) that is really an implementation detail (which gives us the freedom to change it later on if we want).

There are a few items that need to be addressed below:

church89 committed:
addressing Paul's comments

church89 committed:
add renamed files

church89 committed:
add updated result files

church89 committed:
add forgotten docstring

church89 committed:
fix update vec logic

church89 committed:
make keff_search control class private

church89 committed:
update unit test

church89 commented:
Hi @paulromano thanks for the comments. I've addressed now all of them and as suggested I've renamed the class to `KeffSearchControl` and made it private, exposing only the `integrator.add_keff_search_control` method.

church89 committed:
update test results

church89 committed:
update regression tests

church89 committed:
fix bug

paulromano committed:
Fix MPI behavior in _KeffSearchControl._update_vec

paulromano committed:
Style fixes

paulromano committed:
Fix restart with saved keff search

paulromano committed:
Small edits

paulromano committed:
Simplify test

paulromano committed:
Style fixes

paulromano committed:
Refactor Integrator.integrate

paulromano committed:
Remove step_index in search_for_keff calls

paulromano committed:
Remove a deepcopy

paulromano committed:
Rename to run

paulromano committed:
Don't expose keff_search_control

church89 committed:
Merge pull request #55 from paulromano/keff-search-fixes

Fixes for PR 2693

church89 committed:
fix bug during restart

church89 commented:
@paulromano your fixes have been incorporated. Just a small bug during restart that I've noticed after closing your PR in the openmsr branch. 
Also, what is the purpose in the `_restore_keff_search_control` abc method of assigning the restored `keff_search_root` to the function after restart (`self._keff_search_control.function(keff_search_root)` ?

church89 committed:
return null in SIIntegrator

paulromano commented:
When restarting, calling `self._keff_search_control.function(keff_search_root)` will restore the controlled parameter to its saved state (e.g., moving a control rod to the critical position) since the model doesn't otherwise reflect it already.

What was the bug in the restart? I looked at the commit you made but to me the behavior looks the same as before.

church89 commented:
The previous result returned by `self._get_bos_data_from_restart` is an  `OperatorResult` instance, which doesn't contain `keff_search_root'

paulromano commented:
I had deliberately left the keff search control part out of `_get_bos_data_from_restart` because that method is also used in `SIIntegrator.integrate`, where keff search control is not yet supported. For the normal `Integrator` classes, the keff search control was handled immediately after the call to `_get_bos_data_from_restart`. It feels unnatural to me now that `_get_bos_data_from_operator` and `_get_bos_data_from_restart` do not have consistent return types.

church89 committed:
restored previous implementation and replaces OperatorResult with self.operator.res

church89 commented:
`_get_bos_data_from_restart` returns an `OperatorResult` object as result which does not contain `keff_search_root`. I've restored your implementation and just replaced the `OperatorResult` instance with `self.operator.prev_res[-1]` in the call to `_restore_keff_search_control.` Let me know if it makes sense to you now.

paulromano committed:
Add some type hints

paulromano committed:
Add keff_search_root to io_formats doc

paulromano committed:
Increment depletion_results.h5 minor version

church89 committed:
Merge pull request #57 from paulromano/fixes-2693

Final fixes for PR 2693

paulromano reviewed (approved):
Happy to finally get this one merged. Thanks @church89!

church89 commented:
Thanks @paulromano!
